src/System.Net.Http/src/System/Net/Http/SocketsHttpHandler/Http2Connection.cs

```diff
@@ -693,73 +693,86 @@ private async Task SendRstStreamAsync(int streamId, Http2ProtocolErrorCode error
         }
 
-        private void WriteIndexedHeaderName(int index, ReadOnlySpan<byte> value)
+        private void WriteLiteralHeader(string name, string value)
         {
             int bytesWritten;
-            while (!HPackEncoder.EncodeIndexedName(index, value, _headerBuffer.AvailableSpan, out bytesWritten))
+            while (!HPackEncoder.EncodeHeaderNameValue(name, value, _headerBuffer.AvailableSpan, out bytesWritten))
             {
-                _headerBuffer.EnsureAvailableSpace(_headerBuffer.AvailableSpan.Length + 1);
+                _headerBuffer.EnsureAvailableSpace(_headerBuffer.AvailableLength + 1);
             }
 
             _headerBuffer.Commit(bytesWritten);
         }
 
-        private void WriteLiteralHeader(string name, string value)
+        private void WriteBytes(ReadOnlySpan<byte> bytes)
         {
-            int bytesWritten;
-            while (!HPackEncoder.EncodeHeader(name, value, _headerBuffer.AvailableSpan, out bytesWritten))
+            if (_headerBuffer.AvailableLength < bytes.Length)
             {
-                _headerBuffer.EnsureAvailableSpace(_headerBuffer.AvailableSpan.Length + 1);
+                _headerBuffer.EnsureAvailableSpace(bytes.Length);
             }
 
-            _headerBuffer.Commit(bytesWritten);
+            bytes.CopyTo(_headerBuffer.AvailableSpan);
+            _headerBuffer.Commit(bytes.Length);
         }
 
         private void WriteHeaderCollection(HttpHeaders headers)
         {
-            // TODO: ISSUE 31307: Use static table for known headers
-
             foreach (KeyValuePair<HeaderDescriptor, string[]> header in headers.GetHeaderDescriptorsAndValues())
             {
-                // The Host header is not sent for HTTP2 because we send the ":authority" pseudo-header instead
-                // (see pseudo-header handling below in WriteHeaders)
-                if (header.Key.KnownHeader == KnownHeaders.Host)
-                {
-                    continue;
-                }
+                Debug.Assert(header.Value.Length > 0, "No values for header??");
 
-                // The Connection header is not supported in HTTP2. Don't send it.
-                if (header.Key.KnownHeader == KnownHeaders.Connection)
+                KnownHeader knownHeader = header.Key.KnownHeader;
+                if (knownHeader != null)
                 {
-                    continue;
+                    // The Host header is not sent for HTTP2 because we send the ":authority" pseudo-header instead
+                    // (see pseudo-header handling below in WriteHeaders). The Connection header is also not supported
+                    // in HTTP2. Don't send it.
+                    if (knownHeader == KnownHeaders.Host || knownHeader == KnownHeaders.Connection)
+                    {
+                        continue;
+                    }
+
+                    // If there's a static table value for this known header, use that
```

Comment:
An alternative here would be to actually precalculate the generated bytes for the header name. This would save us a small amount of work in constructing these every time. And, it would also work for known headers that aren't in the static table. I don't know if this is actually worth the effort, and it's certainly fine to defer this beyond this PR.

Reply:
You mean basically go back to what I had before, where KnownHeader includes a byte[] for the HTTP2 encoded header, where some are encoded by the associated index and others are encoded as a literal?

Reply:
Yes, basically. The main difference would be that the encoded header is using indexes where possible.

Reply:
Again, I don't think this is that important; I'm happy either way.

Reply:
Ok